boot/freeldr/freeldr/arch/i386/xboxvideo.c

```diff
@@ -122,48 +120,37 @@ XboxVideoPutChar(int Ch, UCHAR Attr, unsigned X, unsigned Y)
   XboxVideoOutputChar(Ch, X, Y, FgColor, BgColor);
 }
 
+UCHAR
+NvGetCrtc(UCHAR Index)
+{
+  *((PUCHAR) NV2A_CRTC_REGISTER_INDEX) = Index;
+  return *((PUCHAR) NV2A_CRTC_REGISTER_VALUE);
+}
+
 VOID
 XboxVideoInit(VOID)
 {
-  ULONG AvMode;
-
   /* Reuse framebuffer that was set up by firmware */
-  FrameBuffer = (PVOID)*((PULONG) 0xfd600800);
+  FrameBuffer = (PVOID)*((PULONG) NV2A_CRTC_FRAMEBUFFER_START);
   /* Verify that framebuffer address is page-aligned */
   ASSERT((ULONG_PTR)FrameBuffer % PAGE_SIZE == 0);
 
   /* FIXME: obtain fb size from firmware somehow (Cromwell reserves high 4 MB of RAM) */
   FrameBufferSize = 4 * 1024 * 1024;
 
-  /* FIXME: don't use SMBus, obtain current video resolution directly from NV2A */
-  if (I2CTransmitByteGetReturn(0x10, 0x04, &AvMode))
-    {
-      if (1 == AvMode) /* HDTV */
-        {
-          ScreenWidth = 720;
-        }
-      else
-        {
-          /* FIXME Other possible values of AvMode:
-           * 0 - AV_SCART_RGB
-           * 2 - AV_VGA_SOG
-           * 4 - AV_SVIDEO
-           * 6 - AV_COMPOSITE
-           * 7 - AV_VGA
-           * other AV_COMPOSITE
-           */
-          ScreenWidth = 640;
-        }
-    }
-  else
-    {
-      ScreenWidth = 640;
-    }
-
-  ScreenHeight = 480;
-  BytesPerPixel = 4;
+  ScreenWidth = *((PULONG) NV2A_RAMDAC_FP_HVALID_END) + 1;
+  ScreenHeight = *((PULONG) NV2A_RAMDAC_FP_VVALID_END) + 1;
+  /* Get BPP directly from NV2A CRTC (magic constants are from Cromwell) */
+  BytesPerPixel = 8 * (((NvGetCrtc(0x19) & 0xE0) << 3) | (NvGetCrtc(0x13) & 0xFF)) / ScreenWidth;
```

Comment:
Here you can do two things in addition:
1. Check whether it's "consistent" with CRTC 0x28 data in case bytesperpixel is 1 or 2 (or 3 ...) (we know that for >= 3 values the value from CRTC 0x28 will always look like 0x83 ...)
2. If BytesPerPixel == 0 then set it to 1 .

Reply:
Added a check for consistency.

> If BytesPerPixel == 0 then set it to 1

Why? I have an assertion here for range 1..4